sw/device/lib/testing/mask_rom_test.h

```diff
@@ -0,0 +1,48 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#include "gmock/gmock.h"
+#include "gtest/gtest.h"
+#include <memory>
+
+namespace mask_rom_test {
+
+/**
+ * Test fixture for mask rom open box tests.
+ *
+ * This class exists to ease mocking free functions in mask rom tests. Test
+ * suites can define test fixtures that inherit from this class by passing the
+ * mock class that they want to use as the template argument `T`. Before a test
+ * starts, this class constructs an instance of this mock class and makes it
+ * available via the static `mock()` method. Free functions can then be mocked
+ * by calling the corresponding methods of the mock class.
+ */
+template <class T>
+class MaskRomOpenBoxTest : public testing::Test {
+ public:
+  static T &mock() { return *mock_; }
+
+ protected:
+  void SetUp() override { mock_ = std::make_unique<testing::StrictMock<T>>(); }
+
+  void TearDown() override { mock_.reset(); }
+
+  thread_local static std::unique_ptr<T> mock_;
```

Comment:
I'm confused... why is this a `thread_local static`, rather than a non-static member variable? There's no guarantee that tests will be executed on distinct threads, so this won't provide any sort of isolation.

Based on the code you've written, I think it makes sense to design mask ROM APIs that we want to be able to mock to pass around some kind of "context" void*, which we can use to inject mocks in tests.

An alternative that might more clearly make this API generic might be something like this:
```cc
template <typename T>
class InjectContext {
 public:
  template<typename Args...>
  InjectContext(T *&global_ptr, Ts&&... args) : place_(place) {
    // Optional: CHECK(global_ptr == nullptr)?
    // You could ostensibly save the old value of global_ptr and swap it
    // back in the dtor.
    place_ = new T(std::forward(args...));
  }
  ~InjectContext() { delete place_; place_ = nullptr; }

  T &operator*() {
    return *place_;
  }
  T *operator->() {
    return place_;
  }

 private:
   T *&place_;
}
```

Then, in your test, you could write
```cc
thread_local MyMock* my_mock;
/// ...
InjectContext ctx(&my_mock, MyMock());
EXPECT_CALL(*ctx, ...);
```

Thus, we can see in each test much more clearly where each mock is injected. You could even imagine wrapping the `thread_local T*` in another type, e.g. `GlobalContext` or something, and give it a member function like `->Inject()` that constructs a new `InjectContext`:
```cc
GlobalContext<MyMock> my_mock;
/// ...
auto ctx = my_mock.Inject(MyMock());
```

What do you think?

Reply:
> I'm confused... why is this a thread_local static, rather than a non-static member variable? There's no guarantee that tests will be executed on distinct threads, so this won't provide any sort of isolation.

This is a `static` because mask rom functions don't have an argument like the DIF functions do that we can use to inject a mock. This is basically what the googletest doc recommends for [sharing resources between tests](https://github.com/google/googletest/blob/master/googletest/docs/advanced.md#sharing-resources-between-tests-in-the-same-test-suite). Regarding `thread_local`, you are right, it is not really needed since AFAIK tests don't run in parallel, this is left over from earlier trials.

> Based on the code you've written, I think it makes sense to design mask ROM APIs that we want to be able to mock to pass around some kind of "context" void*, which we can use to inject mocks in tests.

I would ideally prefer this too but my understanding is that we want to avoid any sort of test-related code in mask rom functions. Apologies for not clearly stating this in the PR. If having this `context` argument and passing it around to every function is acceptable, then we obviously don't need the static member. @gdk, @lenary, any thoughts?

> An alternative that might more clearly make this API generic might be something like this

If we have an argument in all mask rom functions which we can use to inject the mock object in tests, why would we need this? IMHO, just calling the mask rom functions with the mock and using the mock in their definitions in the test file to forward the calls to its methods would be clearer. Is there something that I'm missing here?

Reply:
I'm not very happy with a `void* context` object appearing everywhere in the mask rom, precisely as @mcy points out: it's there for test-only reasons, and we've said we don't want to do that. I think the current approach, as sketched will work for us without needing this.